# Projeto 1 - Ciência dos Dados

Nome: Ana Almeida Barros

Nome: Luana Abramoff

**Atenção:** Serão permitidos grupos de três pessoas, mas com uma rubrica mais exigente. Grupos deste tamanho precisarão fazer um questionário de avaliação de trabalho em equipe

___
Carregando algumas bibliotecas:

In [2]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

**Em `filename`, coloque o nome do seu arquivo de dados!**

In [3]:
import os

filename = 'airpods1.xlsx'
if filename in os.listdir():
    print(f'Encontrei o arquivo {filename}, tudo certo para prosseguir com a prova!')
else:
    print(f'Não encontrei o arquivo {filename} aqui no diretório {os.getcwd()}, será que você não baixou o arquivo?')

Encontrei o arquivo airpods1.xlsx, tudo certo para prosseguir com a prova!


Carregando a base de dados com os tweets classificados como relevantes e não relevantes:

In [4]:
train = pd.read_excel(filename)
train.head(5)

Treinamento  Relevância  Unnamed: 2  \
0  eu não sei se faz muito sentido lançar as airt...           0         NaN   
1  @margaridacs10 @filipa_chaan "comprem os nosso...           0         NaN   
2  apple atualiza airpods pro com suporte a áudio...           0         NaN   
3  @karinnesanso miga, xaomi resolvendo minha vid...           1         NaN   
4  receita federal do aeroporto jfk apreendeu fon...           0         NaN   

    Unnamed: 3  Unnamed: 4   Unnamed: 5  
0          NaN  quantidade  porcentagem  
1    RELEVANTE         132           44  
2  IRRELEVANTE         168           56  
3          NaN         NaN          NaN  
4          NaN         NaN          NaN

In [5]:
test = pd.read_excel(filename, sheet_name = 'Teste')
test.head(5)

Teste  Relevancia  Unnamed: 2  \
0  nao tiro da cabeça ariana grande de wave e air...           0         NaN   
1  airpods pro ganham áudio espacial e troca auto...           1         NaN   
2  alfândega dos eua confunde fones da oneplus co...           0         NaN   
3  putaria do caralho\ncopyright é coisa de retar...           0         NaN   
4  @umameira eles n cabem na minha orelhinha :c\n...           1         NaN   

    Unnamed: 3   Unnamed: 4   Unnamed: 5  
0          NaN  quantidade   porcentagem  
1    RELEVANTE          102      51.7766  
2  IRRELEVANTE           95      48.2234  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN

___
## Classificador automático de sentimento


Faça aqui uma descrição do seu produto e o que considerou como relevante ou não relevante na classificação dos tweets.

Os Airpods são fones de ouvido sem fio da Apple. Na classsificacao dos tweets, buscamos observar palavras, expressoes ou emojis que demonstrassem um senso de opiniao, satisfacao ou emocao em relacao ao produto. Alguns casos vistos foram pessoas declarando que queriam muito o produto, dizendo são muito felizes em tê-lo, fazendo um crítica negativa a algum aspecto pontual e fazendo comparacoes entre marcas e precos. 

___
### Montando um Classificador Naive-Bayes

Considerando apenas as mensagens da planilha Treinamento, ensine  seu classificador.

In [6]:
import re 

def cleanup(text):
    punctuation = '[!-.:?;\n]'
    pattern = re.compile(punctuation)
    text_subbed = re.sub(pattern, '', text)
    return text_subbed

def cleanup2(text):
    punctuation = '[!-.:?;\n]'
    pattern = re.compile(punctuation)
    text_subbed = re.sub(pattern, '', text)
    return text_subbed

In [7]:
relevantes=train[train["Relevância"]==1]
relevantes.head()

Treinamento  Relevância  Unnamed: 2  \
3   @karinnesanso miga, xaomi resolvendo minha vid...           1         NaN   
6   @llllllluyza @maycomsoares_ ao que tudo indica...           1         NaN   
7   agora realmente quero os airpods pro, fuçando ...           1         NaN   
10  apple atualiza firmwares dos airpods (comum e ...           1         NaN   
14  dois motivos me levaram a isso: 1. usar o fone...           1         NaN   

   Unnamed: 3 Unnamed: 4 Unnamed: 5  
3         NaN        NaN        NaN  
6         NaN        NaN        NaN  
7         NaN        NaN        NaN  
10        NaN        NaN        NaN  
14        NaN        NaN        NaN

In [20]:
texto_R=" ".join(relevantes.Treinamento)
texto_R=cleanup(texto_R).lower()
texto_R

'@karinnesanso miga xaomi resolvendo minha vida airpods q tenho a vê 💁🏽\u200d♀️ @llllllluyza @maycomsoares_ ao que tudo indicanovo airpodsanúncio do “novo” processador nos mac’s essas coisas com pouca relevância agora realmente quero os airpods pro fuçando o meu iphone fiz alguma coisa que meu deus que som apple atualiza firmwares dos airpods comum e pro para a versão 3a283 https//tco/ukgah5tliw dois motivos me levaram a isso 1 usar o fone de novo paguei caro nele e tava na gaveta sem usar a meses 2 gasto menos a bateria dos airpods menos ciclos lembro dos airpods pro que eu poderia ter e fico triste serio n sei oq eu quero mais um ipad um airpods ou um mac air eu queria só um airpods de aniversário 🥺 meu airpods n tá prestando 🤡 não gostei desse airpods pro fica feio pra caralho no ouvido prefiro o meu segunda geração mesmo show eu não perdi meus airpods não né quebrei a caixa do meu airpods 🤩 sera que airpods conecta em notebook rt @luanaolimpio_ mermao eu perdi meus airpods real ofi

In [21]:
todas_as_palavras_R=texto_R.split()             
todas_as_palavras_R[0:10]
print(len(todas_as_palavras_R))

2236


In [22]:
serie_train_R = pd.Series(todas_as_palavras_R)
serie_train_R

0       @karinnesanso
1                miga
2               xaomi
3          resolvendo
4               minha
            ...      
2231          airpods
2232             pode
2233               me
2234           deixar
2235            feliz
Length: 2236, dtype: object

In [23]:
tabela_train_R = serie_train_R.value_counts()
tabela_train_R

airpods            130
e                   59
de                  53
que                 50
um                  50
                  ... 
estragou💗🥰😍👏🏻🙏🏻      1
🧘                    1
resolver             1
coração              1
todo                 1
Length: 873, dtype: int64

In [24]:
tabela_train_relativa_R = serie_train_R.value_counts(True)
tabela_train_relativa_R

airpods            0.058140
e                  0.026386
de                 0.023703
que                0.022361
um                 0.022361
                     ...   
estragou💗🥰😍👏🏻🙏🏻    0.000447
🧘                  0.000447
resolver           0.000447
coração            0.000447
todo               0.000447
Length: 873, dtype: float64

In [25]:
irrelevantes=train[train["Relevância"]==0]

texto_I=" ".join(irrelevantes.Treinamento) # um unico texto
texto_I=cleanup(texto_I).lower()

In [26]:
todas_as_palavras_I=texto_I.split()             
todas_as_palavras_I[0:10]

['eu',
 'não',
 'sei',
 'se',
 'faz',
 'muito',
 'sentido',
 'lançar',
 'as',
 'airtags']

In [27]:
serie_train_I = pd.Series(todas_as_palavras_I)
serie_train_I

0                          eu
1                         não
2                         sei
3                          se
4                         faz
                ...          
2790                    piada
2791                      por
2792                 confusão
2793                  confira
2794    https//tco/bshujc7q68
Length: 2795, dtype: object

In [28]:
tabela_train_I = serie_train_I.value_counts()
tabela_train_I #freq absoluta

airpods     154
de           69
que          63
a            52
e            52
           ... 
academia      1
abaixo        1
caixinha      1
oq            1
saiba         1
Length: 1079, dtype: int64

In [29]:
tabela_train_relativa_I = serie_train_I.value_counts(True)
tabela_train_relativa_I

airpods     0.055098
de          0.024687
que         0.022540
a           0.018605
e           0.018605
              ...   
academia    0.000358
abaixo      0.000358
caixinha    0.000358
oq          0.000358
saiba       0.000358
Length: 1079, dtype: float64

Para calcular a chance de um tweet ser relevante ou nao, utiliza-se as seguintes formulas:

$P(R|tweet) = \frac{P(tweet|R)P(R)}{P(tweet)}$

$P(I|tweet) = \frac{P(tweet|I)P(I)}{P(tweet)}$


In [33]:
total_relevantes = 873
total_irrelevantes = 1079 #pelo tamanho das listas de frequência

interseccao = 0


todas_as_palavras_R_semrepetir = []

for i in todas_as_palavras_R:
    if i not in todas_as_palavras_R_semrepetir:
        todas_as_palavras_R_semrepetir.append(i)
        
todas_as_palavras_I_semrepetir = []

for i in todas_as_palavras_I:
    if i not in todas_as_palavras_I_semrepetir:
        todas_as_palavras_I_semrepetir.append(i)

for i in todas_as_palavras_R_semrepetir:
    if i in todas_as_palavras_I_semrepetir:
        interseccao +=1

total_ambas = (total_relevantes + total_irrelevantes) - interseccao
total_ambas

1676

In [41]:
#Definindo
P_R=0.44 
P_I=0.56 #calculadaS na própria planilha do excel

#P_tweet nao precisa ser calculada pois ambas 𝑃(𝑅|𝑡𝑤𝑒𝑒𝑡) e 𝑃(𝐼|𝑡𝑤𝑒𝑒𝑡) sao divididas por ela.


def probabilidadefraseR(x):
    pfrase=1
    palavras = x.split()
    for e in palavras:
        if e not in todas_as_palavras_R_semrepetir:
            ppalavra = 1/(total_relevantes + total_ambas)
        else:
            ppalavra = (tabela_train_R[e]+ 1)/(total_relevantes + total_ambas)
        pfrase = pfrase*ppalavra
    return pfrase

def probabilidadefraseI(x):
    pfrase=1
    palavras = x.split()
    for e in palavras:
        if e not in todas_as_palavras_I_semrepetir:
            ppalavra = 1/(total_relevantes + total_ambas)
        else:
            ppalavra = (tabela_train_I[e]+ 1)/(total_relevantes + total_ambas)
        pfrase = pfrase*ppalavra
    return pfrase

In [42]:
    
for j in test.Teste:
    tweet = j
    j=cleanup(str(j)).lower()
    
    P_tweet_R =probabilidadefraseR(j)
    print(P_tweet_R)
    P_R_tweet=P_tweet_R*P_R
    print(P_R_tweet)
   

    P_tweet_I = probabilidadefraseI(j)
    P_I_tweet = P_tweet_I*P_I
    
    
    if (P_R_tweet> P_I_tweet):
        test.loc[test.Teste==tweet, "Classificador"] = "1"
    else:
        test.loc[test.Teste==tweet, "Classificador"] = "0" 
test

2.0004195214844745e-38
8.801845894531688e-39
1.2787018071649503e-37
5.626287951525781e-38
4.2841975077663275e-38
1.885046903417184e-38
4.9994130588473257e-23
2.1997417458928234e-23
1.679234453717683e-75
7.388631596357806e-76
2.3458551873525313e-33
1.0321762824351138e-33
2.5952315063202755e-36
1.1419018627809212e-36
3.7882722003239327e-57
1.6668397681425304e-57
2.0628444534917852e-46
9.076515595363856e-47
3.792820708483096e-23
1.6688411117325622e-23
1.5511249675699832e-27
6.824949857307926e-28
3.253726329555317e-52
1.4316395850043394e-52
3.59230429231393e-37
1.5806138886181292e-37
6.588566997501815e-23
2.898969478900799e-23
5.355246884707911e-38
2.356308629271481e-38
2.033270485766712e-41
8.946390137373533e-42
4.352142060357604e-74
1.914942506557346e-74
9.502813960350388e-27
4.181238142554171e-27
3.120119842084677e-35
1.372852730517258e-35
1.977984886654912e-18
8.703133501281612e-19
4.258386206291715e-148
1.8736899307683547e-148
1.6245906460030697e-34
7.148198842413507e-35
7.51870878946

Teste  Relevancia  \
0    nao tiro da cabeça ariana grande de wave e air...           0   
1    airpods pro ganham áudio espacial e troca auto...           1   
2    alfândega dos eua confunde fones da oneplus co...           0   
3    putaria do caralho\ncopyright é coisa de retar...           0   
4    @umameira eles n cabem na minha orelhinha :c\n...           1   
..                                                 ...         ...   
192  eu falei mal na sala do airpod da apple pq meu...           1   
193  aquele airpod é 1.000 pila tá louco eu não pag...           1   
194  @analuizatitto amg sim virar cadelinha da appl...           1   
195  como q  eu vou tratar quem tem airpod original...           1   
196  comprei um fone novo de uma marca que chama ha...           0   

     Unnamed: 2   Unnamed: 3   Unnamed: 4   Unnamed: 5 Classificador  
0           NaN          NaN  quantidade   porcentagem             0  
1           NaN    RELEVANTE          102      51.7766             0  
2           NaN  IRRELEVANTE           95      48.2234             0  
3           NaN          NaN          NaN          NaN             0  
4           NaN          NaN          NaN          NaN             1  
..          ...          ...          ...          ...           ...  
192         NaN          NaN          NaN          NaN             0  
193         NaN          NaN          NaN          NaN             1  
194         NaN          NaN          NaN          NaN             1  
195         NaN          NaN          NaN          NaN             0  
196         NaN          NaN          NaN          NaN             0  

[197 rows x 7 columns]

___
### Verificando a performance do Classificador

Agora você deve testar o seu classificador com a base de Testes.

In [45]:
pd.crosstab(test.Relevancia,test.Classificador,normalize=True)

Classificador         0         1
Relevancia                       
0              0.380711  0.101523
1              0.218274  0.299492

38% era irrelevante e foi classificado irrelevante.
29% era relevante e foi classificado relevante.
Assim, 67% dos resultados foram avaliados corretamente.

10% era irrelevante e foi considerado relevante.
21% era relevante e foi considerado irrelevante. 
Assim 31% dos resultados foram avaliados erroneamente. 

___
### Concluindo

___
## Aperfeiçoamento:

Os trabalhos vão evoluir em conceito dependendo da quantidade de itens avançados:

* Limpar: \n, :, ", ', (, ), etc SEM remover emojis
* Corrigir separação de espaços entre palavras e emojis ou entre emojis e emojis
* Propor outras limpezas e transformações que não afetem a qualidade da informação ou classificação
* Criar categorias intermediárias de relevância baseadas na probabilidade: ex.: muito relevante, relevante, neutro, irrelevante, muito irrelevante (3 categorias: C, mais categorias conta para B)
* Explicar por que não posso usar o próprio classificador para gerar mais amostras de treinamento
* Propor diferentes cenários para Naïve Bayes fora do contexto do projeto
* Sugerir e explicar melhorias reais com indicações concretas de como implementar (indicar como fazer e indicar material de pesquisa)
* Montar um dashboard que realiza análise de sentimento e visualiza estes dados

___
## Referências

[Naive Bayes and Text Classification](https://arxiv.org/pdf/1410.5329.pdf)  **Mais completo**

[A practical explanation of a Naive Bayes Classifier](https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/) **Mais simples**